Classes/Property/Service/MarkerParser.php

```diff
@@ -30,12 +31,88 @@ class MarkerParser
      */
     public function replaceMarkers($string, array $markers)
     {
+        if (empty($markers)) {
+            return $string;
+        }
+
+        $markers = $this->keyMarkersByName($markers);
+        $matches = $this->matchMarkers($string);
+
+        if (empty($matches[0])) {
+            return $string;
+        }
+
+        list($identifiers, $variables, $roots) = $matches;
+
         $replacePairs = [];
 
-        foreach ($markers as $marker) {
-            $replacePairs[$marker->getFormattedName()] = $marker->getValue();
+        foreach ($variables as $index => $variable) {
+            $identifier = $identifiers[$index];
+            $root = $roots[$index];
+            $marker = $markers[$root];
+
+            $value = $this->getVariableValue($variable, $root, $marker);
+
+            $replacePairs[$identifier] = $value;
         }
 
         return strtr($string, $replacePairs);
     }
+
+    /**
+     * This will set each marker's name as the array key
+     *
+     * @param array $markers
+     * @return array
+     */
+    private function keyMarkersByName(array $markers)
+    {
+        $aux = [];
+
+        // This will avoid looping on markers for each variable
+        foreach ($markers as $marker) {
+            $aux[$marker->getName()] = $marker;
+        }
+
+        return $aux;
+    }
+
+    /**
+     * This method will find all markers in the string
+     *
+     * @param $string
+     * @return mixed
```

Comment:
Should always return an array. Are we sure the matches are always filled with an array?

Reply:
Yes I tried and when there's no match `$matches` is still an array with the correct structure.